iroh-router/src/router.rs

```diff
@@ -0,0 +1,211 @@
+            let protocols = protos;
+            loop {
+                tokio::select! {
+                    biased;
+                    _ = cancel_token.cancelled() => {
+                        break;
+                    },
+                    // handle incoming p2p connections.
+                    Some(incoming) = endpoint.accept() => {
+                        let protocols = protocols.clone();
+                        join_set.spawn(async move {
+                            handle_connection(incoming, protocols).await;
+                            anyhow::Ok(())
+                        });
+                    },
+                    // handle task terminations and quit on panics.
+                    res = join_set.join_next(), if !join_set.is_empty() => {
+                        match res {
+                            Some(Err(outer)) => {
+                                if outer.is_panic() {
+                                    error!("Task panicked: {outer:?}");
+                                    break;
+                                } else if outer.is_cancelled() {
+                                    debug!("Task cancelled: {outer:?}");
+                                } else {
+                                    error!("Task failed: {outer:?}");
+                                    break;
+                                }
+                            }
+                            Some(Ok(Err(inner))) => {
+                                debug!("Task errored: {inner:?}");
+                            }
+                            _ => {}
+                        }
+                    },
+                    else => break,
+                }
+            }
+
+            shutdown(&endpoint, protocols).await;
+
+            // Abort remaining tasks.
+            tracing::info!("Shutting down remaining tasks");
+            join_set.shutdown().await;
+        };
+        let task = tokio::task::spawn(fut);
+        let task = AbortOnDropHandle::new(task)
+            .map_err(Box::new(|e: JoinError| e.to_string()) as JoinErrToStr)
+            .shared();
+
+        Ok(Router {
+            endpoint: self.endpoint,
+            protocols,
+            task,
+            cancel_token: cancel,
+        })
+    }
+}
+
+/// Shutdown the different parts of the router concurrently.
+async fn shutdown(endpoint: &Endpoint, protocols: Arc<ProtocolMap>) {
+    let error_code = 1u16;
+
+    // We ignore all errors during shutdown.
+    let _ = tokio::join!(
+        // Close the endpoint.
+        // Closing the Endpoint is the equivalent of calling Connection::close on all
+        // connections: Operations will immediately fail with ConnectionError::LocallyClosed.
+        // All streams are interrupted, this is not graceful.
+        endpoint
+            .clone()
+            .close(error_code.into(), b"provider terminating"),
+        // Shutdown protocol handlers.
+        protocols.shutdown(),
```

Comment:
By doing this in a join you're making it a bit non-deterministic whether the protocol handlers will still have an endpoint when their `.shutdown()` is being called.

I'm in two minds on whether the router should be closing the endpoint (given the earlier API/commen of how you pass a clone of the endpoint to the router which makes it seem like you still own it).  If it does however I think it might be best to do this at the very end once all other tasks from this router are stopped.  This would give protocols the possibility of doing a much cleaner shutdown themselves.  Maybe a protocol still wants to exchange some messages with the peer saying "i'm going to close now" or something similar.

Reply:
this is the exact some shutdown routine that we currently have in the `iroh` node, I am okay in changing it if we think this is needed, but lets do this separately please

Reply:
Agreed that we shouldn't change this in this PR, but want to comment on this:
Is this where we should implement a `endpoint.all_closed().await` instead?